src/main/java/com/limechain/trie/structure/TrieStructure.java

```diff
@@ -1,30 +1,480 @@
+     * @return the indices to traverse to reach `target` from root, not including `target` itself.
+     *         So if target is root, returns an empty list.
+     * @throws NullPointerException if targetNodeIndex is not a valid index
+     */
+    List<Integer> nodePath(int targetNodeIndex) {
+        List<Integer> path = new LinkedList<>();
+        var current = this.getNodeAtIndexInner(targetNodeIndex).parent;
+
+        while (current != null) {
+            int nodeIndex = current.parentNodeIndex();
+            path.add(nodeIndex);
+            current = this.getNodeAtIndexInner(nodeIndex).parent;
+        }
+
+        Collections.reverse(path);
+
+        return path;
+
+//        // Alternatively (and more functionally :D):
+//        return Stream.iterate(this.getNodeAtIndexInner(targetNodeIndex).parentIndex,
+//                Objects::nonNull,
+//                index -> this.getNodeAtIndexInner(index.getValue0()).parentIndex)
+//            .map(Pair::getValue0)
+//            .collect(Collectors.collectingAndThen(Collectors.toList(), list -> {
+//                Collections.reverse(list);
+//                return list;
+//            }));
+    }
+
+    /**
+     * Returns true if the structure of this trie is the same as the structure of {@code other}.
+     * Everything is compared for equality except for the {@link TrieNode#userData}s.
+     *
+     * @implNote This method first compares the sizes of the two trie structures,
+     *           and if they don't match, it early returns false.
+     *           Otherwise, it performs the expensive operation of traversing both tries
+     *           in order to identify a potential mismatch.
+     */
+    public <U> boolean structurallyEquals(TrieStructure<U> other) {
+        if (this.nodes.size() != other.nodes.size()) {
+            return false;
+        }
+
+        var thisIter = this.iteratorOrdered();
+        var otherIter = other.iteratorOrdered();
+
+        while (true) {
+            if (!thisIter.hasNext() && !otherIter.hasNext()) {
+                return true;
+            }
+
+            var thisNode = this.getNodeAtIndexInner(thisIter.next().getValue());
+            var otherNode = other.getNodeAtIndexInner(otherIter.next().getValue());
+
+            if (thisNode.hasStorageValue != otherNode.hasStorageValue) {
+                return false;
+            }
+
+            // Check if the parents match
+            {
+                var thisNodeParent = thisNode.parent;
+                var otherNodeParent = otherNode.parent;
+
+                //NOTE: this is awful, so a quick explainer:
+                //  We want to return false in all cases except:
+                //    - both parents are null;
+                //    - both parents are not null and the two nodes' child indices within them are the same
+                //  Reference: https://github.com/smol-dot/smoldot/blob/200214a571af30b5fa3997aea988451adc235ed0/lib/src/trie/trie_structure.rs#L615
```

Comment:
I'm open for suggestions here, that's really ugly but I have no better ideas.

Reply:
Just split it up a bit into more variables and it will become very easy to read:
```
boolean areBothNull = ...
boolean areBothNotNullWithSameChildIndex = ...
if (areBothNull || areBothNotNullWithSameChildIndex) doSmth() else false
```

Reply:
Good one. I've still left the clarifying comment above the block, I reckoned a human language explanation could still be helpful.